File: molview/src/chem/rings.ts
import type { Molecule, Ring } from './types';
import { neighborLists } from './smiles';

/**
 * Поиск наименьшего набора наименьших циклов (SSSR).
 *
 * Приём простой и надёжный: для каждой связи мысленно её удаляем и ищем
 * кратчайший путь между её концами. Путь плюс сама связь и есть цикл,
 * причём наименьший из проходящих через эту связь. Дальше отбираем
 * независимые циклы — столько, сколько даёт цикломатическое число графа.
 */
export function findRings(mol: Molecule): Ring[] {
  const nb = neighborLists(mol);
  const heavy = mol.atoms.filter((a) => a.el !== 'H').map((a) => a.id);
  const isHeavy = new Set(heavy);

  const candidates: number[][] = [];

  for (const bond of mol.bonds) {
    if (!isHeavy.has(bond.a) || !isHeavy.has(bond.b)) continue;
    const path = shortestPath(nb, isHeavy, bond.a, bond.b, bond);
    if (path && path.length >= 3) candidates.push(path);
  }

  // уникальные циклы по набору атомов, от меньших к большим
  const seen = new Set<string>();
  const unique: number[][] = [];
  for (const c of candidates.sort((x, y) => x.length - y.length)) {
    const key = [...c].sort((a, b) => a - b).join(',');
    if (seen.has(key)) continue;
    seen.add(key);
    unique.push(c);
  }

  // цикломатическое число: сколько независимых циклов есть в графе
  const heavyBonds = mol.bonds.filter((b) => isHeavy.has(b.a) && isHeavy.has(b.b));
  const components = countComponents(nb, heavy, isHeavy);
  const cycleCount = heavyBonds.length - heavy.length + components;

  const chosen: number[][] = [];
  const coveredBonds = new Set<string>();
  for (const c of unique) {
    if (chosen.length >= cycleCount) break;
    const bondKeys = ringBondKeys(c);
    if (bondKeys.some((k) => !coveredBonds.has(k)) || chosen.length === 0) {
      chosen.push(c);
      bondKeys.forEach((k) => coveredBonds.add(k));
    }
  }

  return chosen.map((atoms) => describeRing(mol, atoms));
}

function ringBondKeys(ring: number[]): string[] {
  const keys: string[] = [];
  for (let i = 0; i < ring.length; i++) {
    const a = ring[i];
    const b = ring[(i + 1) % ring.length];
    keys.push(a < b ? `${a}-${b}` : `${b}-${a}`);
  }
  return keys;
}

function shortestPath(
  nb: number[][],
  isHeavy: Set<number>,
  from: number,
  to: number,
  forbidden: { a: number; b: number },
): number[] | null {
  const prev = new Map<number, number>();
  const queue = [from];
  const visited = new Set<number>([from]);

  while (queue.length > 0) {
    const cur = queue.shift()!;
    if (cur === to) break;
    for (const next of nb[cur]) {
      if (!isHeavy.has(next)) continue;
      const isForbidden =
        (cur === forbidden.a && next === forbidden.b) || (cur === forbidden.b && next === forbidden.a);
      if (isForbidden) continue;
      if (visited.has(next)) continue;
      visited.add(next);
      prev.set(next, cur);
      queue.push(next);
    }
  }

  if (!visited.has(to)) return null;
  const path: number[] = [];
  let cur: number | undefined = to;
  while (cur !== undefined) {
    path.push(cur);
    if (cur === from) break;
    cur = prev.get(cur);
  }
  return path.length >= 3 ? path.reverse() : null;
}

function countComponents(nb: number[][], heavy: number[], isHeavy: Set<number>): number {
  const visited = new Set<number>();
  let count = 0;
  for (const start of heavy) {
    if (visited.has(start)) continue;
    count++;
    const stack = [start];
    visited.add(start);
    while (stack.length > 0) {
      const cur = stack.pop()!;
      for (const next of nb[cur]) {
        if (!isHeavy.has(next) || visited.has(next)) continue;
        visited.add(next);
        stack.push(next);
      }
    }
  }
  return count;
}

function describeRing(mol: Molecule, atoms: number[]): Ring {
  let aromatic = true;
  for (let i = 0; i < atoms.length; i++) {
    const a = atoms[i];
    const b = atoms[(i + 1) % atoms.length];
    const bond = mol.bonds.find((x) => (x.a === a && x.b === b) || (x.a === b && x.b === a));
    if (!bond || !bond.aromatic) { aromatic = false; break; }
  }
  return { atoms, aromatic, planar: aromatic };
}

/** Группировка циклов в конденсированные системы (имеющие общие атомы). */
export function fuseRingSystems(rings: Ring[]): Ring[][] {
  const groups: Ring[][] = [];
  const assigned = new Array(rings.length).fill(false);

  for (let i = 0; i < rings.length; i++) {
    if (assigned[i]) continue;
    const group = [rings[i]];
    assigned[i] = true;
    let grew = true;
    while (grew) {
      grew = false;
      const inGroup = new Set<number>();
      group.forEach((r) => r.atoms.forEach((a) => inGroup.add(a)));
      for (let j = 0; j < rings.length; j++) {
        if (assigned[j]) continue;
        if (rings[j].atoms.some((a) => inGroup.has(a))) {
          group.push(rings[j]);
          assigned[j] = true;
          grew = true;
        }
      }
    }
    groups.push(group);
  }
  return groups;
}

File: molviewcpp/src/chem/rings.cpp
#include "chem/rings.h"

#include <algorithm>
#include <deque>
#include <map>
#include <set>
#include <string>

#include "chem/smiles.h"

namespace chem {

namespace {

/** Ключ неупорядоченной пары атомов. */
std::string pairKey(int a, int b) {
    return a < b ? std::to_string(a) + "-" + std::to_string(b)
                 : std::to_string(b) + "-" + std::to_string(a);
}

std::vector<std::string> ringBondKeys(const std::vector<int>& ring) {
    std::vector<std::string> keys;
    for (std::size_t i = 0; i < ring.size(); i++) {
        keys.push_back(pairKey(ring[i], ring[(i + 1) % ring.size()]));
    }
    return keys;
}

/** Кратчайший путь из from в to обходом в ширину, минуя запрещённую связь. */
std::vector<int> shortestPath(const std::vector<std::vector<int>>& nb,
                              const std::set<int>& isHeavy,
                              int from, int to, const Bond& forbidden) {
    std::map<int, int> prev;
    std::deque<int> queue{from};
    std::set<int> visited{from};

    while (!queue.empty()) {
        const int cur = queue.front();
        queue.pop_front();
        if (cur == to) break;
        for (int next : nb[cur]) {
            if (isHeavy.count(next) == 0) continue;
            const bool isForbidden =
                (cur == forbidden.a && next == forbidden.b) || (cur == forbidden.b && next == forbidden.a);
            if (isForbidden) continue;
            if (visited.count(next) != 0) continue;
            visited.insert(next);
            prev.emplace(next, cur);
            queue.push_back(next);
        }
    }

    if (visited.count(to) == 0) return {};

    std::vector<int> path;
    int cur = to;
    for (;;) {
        path.push_back(cur);
        if (cur == from) break;
        const auto it = prev.find(cur);
        if (it == prev.end()) break;
        cur = it->second;
    }
    if (path.size() < 3) return {};
    std::reverse(path.begin(), path.end());
    return path;
}

int countComponents(const std::vector<std::vector<int>>& nb, const std::vector<int>& heavy,
                    const std::set<int>& isHeavy) {
    std::set<int> visited;
    int count = 0;
    for (int start : heavy) {
        if (visited.count(start) != 0) continue;
        count++;
        std::vector<int> stack{start};
        visited.insert(start);
        while (!stack.empty()) {
            const int cur = stack.back();
            stack.pop_back();
            for (int next : nb[cur]) {
                if (isHeavy.count(next) == 0 || visited.count(next) != 0) continue;
                visited.insert(next);
                stack.push_back(next);
            }
        }
    }
    return count;
}

Ring describeRing(const Molecule& mol, const std::vector<int>& atoms) {
    bool aromatic = true;
    for (std::size_t i = 0; i < atoms.size(); i++) {
        const Bond* bond = findBond(mol, atoms[i], atoms[(i + 1) % atoms.size()]);
        if (bond == nullptr || !bond->aromatic) { aromatic = false; break; }
    }
    Ring ring;
    ring.atoms = atoms;
    ring.aromatic = aromatic;
    ring.planar = aromatic;
    return ring;
}

}  // namespace

std::vector<Ring> findRings(const Molecule& mol) {
    const std::vector<std::vector<int>> nb = neighborLists(mol);

    std::vector<int> heavy;
    std::set<int> isHeavy;
    for (const Atom& a : mol.atoms) {
        if (a.el != "H") { heavy.push_back(a.id); isHeavy.insert(a.id); }
    }

    std::vector<std::vector<int>> candidates;
    for (const Bond& bond : mol.bonds) {
        if (isHeavy.count(bond.a) == 0 || isHeavy.count(bond.b) == 0) continue;
        const std::vector<int> path = shortestPath(nb, isHeavy, bond.a, bond.b, bond);
        if (path.size() >= 3) candidates.push_back(path);
    }

    // уникальные циклы по набору атомов, от меньших к большим;
    // сортировка устойчивая — при равной длине порядок задан порядком связей
    std::stable_sort(candidates.begin(), candidates.end(),
                     [](const std::vector<int>& x, const std::vector<int>& y) {
                         return x.size() < y.size();
                     });

    std::set<std::string> seen;
    std::vector<std::vector<int>> unique;
    for (const std::vector<int>& c : candidates) {
        std::vector<int> sorted = c;
        std::sort(sorted.begin(), sorted.end());
        std::string key;
        for (std::size_t k = 0; k < sorted.size(); k++) {
            if (k > 0) key += ',';
            key += std::to_string(sorted[k]);
        }
        if (seen.count(key) != 0) continue;
        seen.insert(key);
        unique.push_back(c);
    }

    // цикломатическое число: сколько независимых циклов есть в графе
    int heavyBondCount = 0;
    for (const Bond& b : mol.bonds) {
        if (isHeavy.count(b.a) != 0 && isHeavy.count(b.b) != 0) heavyBondCount++;
    }
    const int components = countComponents(nb, heavy, isHeavy);
    const int cycleCount = heavyBondCount - static_cast<int>(heavy.size()) + components;

    std::vector<std::vector<int>> chosen;
    std::set<std::string> coveredBonds;
    for (const std::vector<int>& c : unique) {
        if (static_cast<int>(chosen.size()) >= cycleCount) break;
        const std::vector<std::string> bondKeys = ringBondKeys(c);
        bool hasNew = false;
        for (const std::string& k : bondKeys) {
            if (coveredBonds.count(k) == 0) { hasNew = true; break; }
        }
        if (hasNew || chosen.empty()) {
            chosen.push_back(c);
            for (const std::string& k : bondKeys) coveredBonds.insert(k);
        }
    }

    std::vector<Ring> result;
    for (const std::vector<int>& atoms : chosen) result.push_back(describeRing(mol, atoms));
    return result;
}

std::vector<std::vector<Ring>> fuseRingSystems(const std::vector<Ring>& rings) {
    std::vector<std::vector<Ring>> groups;
    std::vector<bool> assigned(rings.size(), false);

    for (std::size_t i = 0; i < rings.size(); i++) {
        if (assigned[i]) continue;
        std::vector<Ring> group{rings[i]};
        assigned[i] = true;
        bool grew = true;
        while (grew) {
            grew = false;
            std::set<int> inGroup;
            for (const Ring& r : group) {
                for (int a : r.atoms) inGroup.insert(a);
            }
            for (std::size_t j = 0; j < rings.size(); j++) {
                if (assigned[j]) continue;
                bool shares = false;
                for (int a : rings[j].atoms) {
                    if (inGroup.count(a) != 0) { shares = true; break; }
                }
                if (shares) {
                    group.push_back(rings[j]);
                    assigned[j] = true;
                    grew = true;
                }
            }
        }
        groups.push_back(group);
    }
    return groups;
}

}  // namespace chem
